go/private/mode.bzl

```diff
@@ -80,8 +80,8 @@ def get_mode(ctx, go_toolchain, cgo_context_info, go_config_info):
     debug = go_config_info.debug if go_config_info else False
     linkmode = go_config_info.linkmode if go_config_info else LINKMODE_NORMAL
     cover_format = go_config_info and go_config_info.cover_format
-    goos = go_toolchain.default_goos
-    goarch = go_toolchain.default_goarch
+    goos = go_toolchain.default_goos if getattr(ctx.attr, "goos", "auto") == "auto" else ctx.attr.goos
```

Comment:
I'm slightly surprised that this line had to be modified. Could you briefly explain why?

Reply:
Yes! It's this stack of calls, if `goos` is explicitly set on a `go_binary`:
`_go_binary_impl` -> `emit_binary` -> `emit_archive`

In `emit_archive` for the binary, we have this check: `if a.source.mode != go.mode: fail`.

If `goos` is set, `go.mode` is set to that `goos`, but currently the `a.source` uses the overriden `goos`, but `go.mode` uses the platform default (which makes sense - the toolchain is still an exec-native toolchain, we're just telling it to build targeting a different `goos`).

This causes an error, e.g. when building `//tests/core/cross:linux_cross` on a mac, because `go.mode` is set to `darwin_amd64_pure`, causing the check in `emit_archive` to fail.

Reply:
That makes sense. Can you point to the part of this PR that makes this change necessary? Why do the tests fail only now?

Reply:
I believe it's the removal of `go_transition_rule` - specifically of setting `cfg = go_transition` in `go_transition_rule`. Before the rule would get transitioned, so would get a new platform:

https://github.com/bazelbuild/rules_go/blob/52a7c74b3975bf9833e4c223370375b8b798a412/go/private/rules/transition.bzl#L177-L187

which would in turn pick a new toolchain, and so have the correct `default_goos`.

Now, instead, we need to decide the goos based on the target, rather than the self-transition, because the self-transition doesn't happen any more.

I _think_ this is the only place that we read these settings assuming they're transitioned, and the tests agree, but it's for sure possible I've missed some others that we will discover :)